### PR TITLE
Barchart gene expression

### PR DESCRIPTION
**Pull request by airenzp**
## Description

In this PR multiple fixes and improvements are done.  The geneExpression tab is supported from the dictionary and summary plots. The method trigger_getGeneExpViolinPlotData was removed and the default backend violin is used instead, since once we support reading the sample values in getData the logic is the same for building the violin. The bins logic was also improved, the bins are initialized from fillTermWrapper only as this ensures they are always set on the server side. The active bins are set in the byTermId bins from getData, which allows to use them to order the scatter legend. In common another helper function was added and used, isDictionaryTerm, to improve readability. Please check if everything is working from the barchart and violin when using geneExpression terms, etc. For testing you can use TermdbTest with TP53 or even the panallMds3 dataset. This PR also fixes some bugs found when testing.

## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

File: client/mass/groups.js [hunk 1]
```diff
@@ -506,7 +508,7 @@ export async function openSummaryPlot(term, samplelstTW, app, id, newId) {
 		config
 	})
 }
-
+let tip2
```
Comment (xzhou82) [inline client/mass/groups.js:511]: please comment on what is tip2

Comment (airenzp) [reply client/mass/groups.js]: Done

File: client/plots/barchart.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ import { renderTable } from '../dom/table'
 import { fillTermWrapper } from '#termsetting'
 import { getColors } from '#shared/common'
 import { mclass } from '#shared/common'
-import { plotColor } from '../shared/common'
+import { isNumericTerm, plotColor } from '../shared/common'
```
Comment (xzhou82) [inline client/plots/barchart.js:16]: should always import from `#shared/common` also can combine line 14-16 into one import

File: client/plots/regression.inputs.term.js [hunk 3]
```diff
@@ -1,12 +1,12 @@
 import { termsettingInit } from '#termsetting'
-import { nonDictionaryTermTypes } from '#shared/termdb.usecase'
 import { getNormalRoot } from '#filter'
 import { get_bin_label } from '#shared/termdb.bins'
 import { InputValuesTable } from './regression.inputs.values.table'
 import { Menu } from '#dom/menu'
 import { select } from 'd3-selection'
 import { mayRunSnplstTask } from '../termsetting/handlers/snplst.sampleSum.ts'
 import { get_defaultQ4fillTW } from './regression'
+import { isDictionaryType } from '../shared/common.js'
```
Comment (xzhou82) [inline client/plots/regression.inputs.term.js:9]: import as `#shared/common`

File: client/termdb/TermTypeSearch.ts [hunk 4]
```diff
@@ -15,9 +15,9 @@ The tree target is used to determine the allowed term types.
 const useCases = {
 	matrix: [TermTypeGroups.DICTIONARY_VARIABLES, TermTypeGroups.MUTATION_CNV_FUSION],
 	filter: [TermTypeGroups.DICTIONARY_VARIABLES, TermTypeGroups.MUTATION_CNV_FUSION],
-	dictionary: [TermTypeGroups.DICTIONARY_VARIABLES, TermTypeGroups.MUTATION_CNV_FUSION],
-	summary: [TermTypeGroups.DICTIONARY_VARIABLES],
-	barchart: [TermTypeGroups.DICTIONARY_VARIABLES, TermTypeGroups.MUTATION_CNV_FUSION],
+	dictionary: [TermTypeGroups.DICTIONARY_VARIABLES, TermTypeGroups.MUTATION_CNV_FUSION, TermTypeGroups.GENE_EXPRESSION],
+	summary: [TermTypeGroups.DICTIONARY_VARIABLES, TermTypeGroups.GENE_EXPRESSION],
+	barchart: [TermTypeGroups.DICTIONARY_VARIABLES, TermTypeGroups.MUTATION_CNV_FUSION, TermTypeGroups.GENE_EXPRESSION],
 	violin: [TermTypeGroups.DICTIONARY_VARIABLES, TermTypeGroups.MUTATION_CNV_FUSION],
```
Comment (xzhou82) [inline client/termdb/TermTypeSearch.ts:21]: add geneexp for violin

File: client/plots/barchart.js (hunk 2 above)
Comment (airenzp) [reply client/plots/barchart.js]: Done

**Comment by xzhou82** [comment]
i added [this violin link](http://localhost:3000/?noheader=1&mass={%22dslabel%22:%22TermdbTest%22,%22genome%22:%22hg38-test%22,%22nav%22:{%22activeTab%22:1},%22plots%22:[{%22chartType%22:%22summary%22,%22term%22:{%22term%22:{%22type%22:%22geneExpression%22,%22gene%22:%22TP53%22,%22name%22:%22TP53%22,%22id%22:%22TP53%22},%22q%22:{%22mode%22:%22continuous%22}}}]}) to url.html (please pull sjpp). on changing the gene to discrete and use `Automatic` for last bin start, it breaks

File: client/termdb/Vocab.js [hunk 5]
```diff
@@ -143,16 +144,19 @@ export class Vocab {
...
 		const copy = { $id: tw.$id, term: {}, q: tw.q }
-		if (tw.term?.type == 'geneVariant') {
+		if (!isDictionaryType(tw.term?.type)) {
+			if (tw.term.id) copy.term.id = tw.term.id
 			copy.term.name = tw.term.name
 			copy.term.type = tw.term.type
 			if (tw.term.gene) {
 				copy.term.gene = tw.term.gene
-			} else {
+			} else if (tw.term.type.startsWith('gene')) {
```
Comment (xzhou82) [inline client/termdb/Vocab.js:153]: please mark this as quick fix and will be fixed later. consider dna methylation

File: server/src/termdb.js [hunk 6]
```diff
@@ -463,3 +464,24 @@ async function LDoverlay(q, ds, res) {
...
+		genome: q.genome,
+		dslabel: q.dslabel,
+		clusterMethod: 'hierarchical',
+		/** distance method */
+		distanceMethod: 'euclidean',
+		/** Data type */
+		dataType: 3,
+		genes: [{ gene: tw.term.gene }]
+	}
+	const data = await ds.queries.geneExpression.get(args)
```
Comment (xzhou82) [inline server/src/termdb.js:480]: fix order, only query data when not cached

File: server/src/termdb.js [hunk 7]
```diff
@@ -463,3 +464,24 @@ async function LDoverlay(q, ds, res) {
...
+	let defaultBins
+	if (ds.queries.geneExpression.gene2bins[tw.term.gene])
+		defaultBins = { default: ds.queries.geneExpression.gene2bins[tw.term.gene] }
+	else defaultBins = { default: getDefaultGeneExpBins(ds, tw, data) }
```
Comment (xzhou82) [inline server/src/termdb.js:484]: convert `data` into array of numbers and call `initBinConfig()`

Comment (airenzp) [reply server/src/termdb.js]: solved now

File: server/src/termdb.js (hunk 6 above)
Comment (airenzp) [reply server/src/termdb.js:486]: done

**Review by xzhou82 — APPROVED** [review]
minor issues:
at [geneexp violin](http://localhost:3000/?noheader=1&mass={%22dslabel%22:%22TermdbTest%22,%22genome%22:%22hg38-test%22,%22nav%22:{%22activeTab%22:1},%22plots%22:[{%22chartType%22:%22summary%22,%22term%22:{%22term%22:{%22type%22:%22geneExpression%22,%22gene%22:%22TP53%22,%22name%22:%22TP53%22,%22id%22:%22TP53%22},%22q%22:{%22mode%22:%22continuous%22}}}]}), select a range and List samples, the listing includes all samples but does not follow limit

in same gene exp violin as above, first switch to discrete/custom bin, then try to switch to regular bin but barchart stays as custom bin

(unrelated) lasso and list on [dynamic scatter](http://localhost:3000/?noheader=1&mass={%22dslabel%22:%22SJLife%22,%22genome%22:%22hg38%22,%22nav%22:{%22activeTab%22:1},%22plots%22:[{%22chartType%22:%22summary%22,%22term%22:{%22id%22:%22eur_sjlife_pc1%22,%22q%22:{%22mode%22:%22continuous%22}},%22term2%22:{%22id%22:%22eur_sjlife_pc2%22,%22q%22:{%22mode%22:%22continuous%22}}}]}) breaks

**Pull request merged**